main.cpp

```diff
@@ -36,26 +36,21 @@ struct T
 
 struct Comparison                                //4
 {
-    T* compare(T* a, T* b) //5
+    T* compare(T& a, T& b) //5
     {
-        if (a != nullptr && b != nullptr)
-        {
-            if( a->value < b->value ) return a;
-            if( a->value > b->value ) return b;
-        }
-        return nullptr;
+        if ((a.value - b.value) < 0.00001) return nullptr;
+        if( a.value < b.value ) return &a;
+        return &b;
```

Comment:
THe original logic was fine.  the nullptr checks just needed to be removed. 
if a < b, return a.
if a > b, return b
return nullptr.

it's simple and easy to read. 
- [ ] revise